### PR TITLE
【Master】【Bug】Fix a bug,  When the worker service offline, workerNodeInfo cache in master cannot delete the offline worker

### LINKED ISSUE #15370: [Bug] [Master] When a worker offline, the master still tries to dispatch tasks to the worker and throws an exception
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

When a worker offline, there is a chance that the master will continue to send tasks to this worker and throw an exception:

[ERROR] 2023-12-20 22:58:25.317 +0800 org.apache.dolphinscheduler.server.master.runner.GlobalTaskDispatchWaitingQueueLooper:[87] - [WorkflowInstance-0][TaskInstance-0] - Dispatch task failed
org.apache.dolphinscheduler.server.master.exception.TaskDispatchException: Dispatch task to 192.168.1.128:1234 failed
        at org.apache.dolphinscheduler.server.master.runner.BaseTaskDispatcher.doDispatch(BaseTaskDispatcher.java:101)
        at org.apache.dolphinscheduler.server.master.runner.BaseTaskDispatcher.dispatchTask(BaseTaskDispatcher.java:74)
        at org.apache.dolphinscheduler.server.master.runner.GlobalTaskDispatchWaitingQueueLooper.run(GlobalTaskDispatchWaitingQueueLooper.java:79)
Caused by: org.apache.dolphinscheduler.remote.exceptions.RemotingException: connect to : Host(ip=192.168.1.128, port=1234) fail
        at org.apache.dolphinscheduler.remote.NettyRemotingClient.sendSync(NettyRemotingClient.java:210)
        at org.apache.dolphinscheduler.server.master.rpc.MasterRpcClient.sendSyncCommand(MasterRpcClient.java:49)
        at org.apache.dolphinscheduler.server.master.runner.BaseTaskDispatcher.doDispatch(BaseTaskDispatcher.java:87)

Snipaste_2023-12-27_19-11-47

### What you expected to happen

When the worker is offline, clear the worker node information in the master and do not attempt to send tasks to it

### How to reproduce

1. Run master-1 and worker-1 server
2. Run worker-2 server. All worker services are in the default worker group
3. kill worker-1 server
4. Running a workflow consists of a number of tasks
5. Some exceptions will be printed to the master log

in addition, 
we can debug master server at `org.apache.dolphinscheduler.server.master.registry.ServerNodeManager#updateWorkerNodes()`, We can find that when a worker is offline, the master does not clear the offline worker information in `workerNodeInfo`

### Anything else

_No response_

### Version

3.2.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by shilei-ly**
## Purpose of the pull request

Fix #15370  
When the worker service goes offline, delete the worker information from the `workerNodeInfo` cache in master

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManager.java [hunk 1]
```diff
@@ -193,6 +194,16 @@ private void syncSingleWorkerNodeInfo(String workerAddress, WorkerHeartBeat info
                 workerNodeInfoWriteLock.unlock();
             }
         }
+
+        private void removeSingleWorkerNode(String workerAddress) {
+            workerNodeInfoWriteLock.lock();
+            try {
+                workerNodeInfo.remove(workerAddress);
+                log.info("remove work {} from workerNodeInfo when work server down", workerAddress);
```
Comment (SbloodyS) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManager.java]:
```suggestion
                log.info("remove worker node {} from workerNodeInfo when worker server down", workerAddress);
```

**Comment by shilei-ly** [comment]
@SbloodyS  Hi，Do you think there are any other problem with this PR

**Comment by SbloodyS** [comment]
Could you add some UT for it? @alei1206

**Comment by shilei-ly** [comment]
> Could you add some UT for it? @alei1206

@SbloodyS  Hi, Unit tests have been added.

**Comment by shilei-ly** [comment]
> > Could you add some UT for it? @alei1206
> 
> @SbloodyS Hi, Unit tests have been added.

@SbloodyS Hi, do you have other suggestions about this code?

**Comment by rickchengx** [comment]
HI, @alei1206 please Run 'mvn spotless:apply' to fix the code style

**Comment by shilei-ly** [comment]
> HI, @alei1206 please Run 'mvn spotless:apply' to fix the code style

@rickchengx  Hi，I have modified the code style using `mvn spotless:apply`. Please check again.

File: dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package org.apache.dolphinscheduler.server.master.registry;
```
Comment (rickchengx) [inline dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManagerTest.java:18]: please add license header

Comment (shilei-ly) [reply dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManagerTest.java:18]: Thank you for reminding, the License Header has been added

**Review by rickchengx — APPROVED** [review]

**Comment by ruanwenjun** [comment]
Does anyone see the `todo` here? Why only handle the delete event, in fact this is not a bug, the task dispatch failed it will retry until the master refresh worker node, in this PR only handle the delete event, but still miss add event is not a good pr.

**Comment by shilei-ly** [comment]
> Does anyone see the `todo` here? Why only handle the delete event, in fact this is not a bug, the task dispatch failed it will retry until the master refresh worker node, in this PR only handle the delete event, but still miss add event is not a good pr.

@ruanwenjun Hi, I think,  when master refresh worker node， only add new worker server to cache,  but not delete the offline worker。 if we not handle the delete event, the offline worker server are always in the cache.   Additionally, `updateWorkerNodes()` will get all the worker nodes in zk and add them to the cache, so I think, we shouldn't need to handle the add event.
image

**Pull request merged**